### PR TITLE
Raise `ArgumentError` when calling `assert_redirected_to` with invalid `:status`

### PR DESCRIPTION
**Pull request by seanpdoyle**
### Motivation / Background

Prior to this commit, it was possible for test suites to misuse the `assert_redirected_to` assertion by passing a `:status` option for codes outside the `3XX` range.

### Detail

This change raises an `ArgumentError` to guide callers toward a redirect code.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Changes that are unrelated should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

File: actionpack/lib/action_dispatch/testing/assertions/response.rb [hunk 1]
```diff
@@ -56,8 +56,12 @@ def assert_response(type, message = nil)
       #   assert_redirected_to "/some/path", status: :moved_permanently
       def assert_redirected_to(url_options = {}, options = {}, message = nil)
         options, message = {}, options unless options.is_a?(Hash)
-
         status = options[:status] || :redirect
+
+        if AssertionResponse.new(status).code in ...299 | 400... | /\A[1245]/
```
Comment (jonathanhefner) [inline actionpack/lib/action_dispatch/testing/assertions/response.rb:61]: Pattern matching is pretty cool, but I think we could simplify this:

```suggestion
        unless AssertionResponse.new(status).code.to_s.match?(/\A3..\z/)
```

Or another possibility would be to put part of this logic in a `AssertionResponse#code_group` method.  Then we could write something like:

```suggestion
        unless AssertionResponse.new(status).code_group == "3XX"
```

Though I'm not sure if there are additional use cases to justify that.

File: actionpack/lib/action_dispatch/testing/assertions/response.rb [hunk 2]
```diff
@@ -56,8 +56,12 @@ def assert_response(type, message = nil)
...
+          raise ArgumentError.new("Expected :status to be a redirect, but was #{status}")
```
Comment (jonathanhefner) [inline actionpack/lib/action_dispatch/testing/assertions/response.rb:62]:
```suggestion
          raise ArgumentError, "Expected :status to be for a redirect, but was #{status}"
```

File: actionpack/test/controller/action_pack_assertions_test.rb [hunk 3]
```diff
@@ -470,6 +470,31 @@ def test_assert_redirection_with_status
...
+    end
+
+    ActionDispatch::AssertionResponse::GENERIC_RESPONSE_CODES.except(:redirect).each do |status, pseudostatus|
+      assert_raise ArgumentError, match: "Expected :status to be a redirect, but was #{status}" do
+        assert_redirected_to("http://test.host/some/path", status: status)
+      end
+      assert_raise ArgumentError, match: "Expected :status to be a redirect, but was #{pseudostatus}" do
+        assert_redirected_to("http://test.host/some/path", status: pseudostatus)
+      end
+    end
```
Comment (jonathanhefner) [inline actionpack/test/controller/action_pack_assertions_test.rb:495]: I think we can omit these assertions.  Based on

https://github.com/rails/rails/blob/1e4f16108401da9952bddf9333b99181f57fe631/actionpack/lib/action_dispatch/testing/assertions/response.rb#L33-L37

`assert_response` does not support arguments like "4XX", and `ActionDispatch::Assertions::RESPONSE_PREDICATES.keys` necessarily covers `ActionDispatch::AssertionResponse::GENERIC_RESPONSE_CODES.keys`.

**Comment by rafaelfranca** [comment]
Thank you for the pull request. Rails never protected users from their own mistakes. If they write `assert_redirected_to "/", status: 200` and they actually return a 200 response and think that is a redirect, isn't the job of the framework to prevent that.

**Pull request closed without merging**